Update MQTT cover template handling (#17738)
home-assistant/home-assistant.github.io: source/_integrations/cover.mqtt.markdown

  type: integer
  default: 100
position_template:

>   description: "Defines a [template](/topics/templating/) that can be used to extract the payload for the `position_topic` topic. Within the template the following variables are available for the calculation of the position: __entity_id__, The ID of the entity itself. It can be used to reverence its attribuetes; __position_open__; __position_closed__; __tilt_min__; __tilt_max__"

Reviewer: ```suggestion
  description: "Defines a [template](/topics/templating/) that can be used to extract the payload for the `position_topic` topic. Within the template the following variables are available for the calculation of the position: `entity_id`, the ID of the entity itself. The `entity_id` can be used to reference the entity's attributes; `position_open`; `position_closed`; `tilt_min`; `tilt_max`; with help of the [states](/docs/configuration/templating/#states) template function."
```

Reviewer: Maybe drop the list of attributes here and just mention that in the "Full configuration using advanced templating" section.

>   description: "Defines a [template](/topics/templating/) that can be used to extract the payload for the `position_topic` topic. Within the template the following variables are available for the calculation of the position: __entity_id__, The ID of the entity itself. It can be used to reverence its attribuetes; __position_open__; __position_closed__; __tilt_min__; __tilt_max__"

Reviewer: ```suggestion
  description: "Defines a [template](/topics/templating/) that can be used to extract the payload for the `position_topic` topic. Within the template the following variables are available: `entity_id`, `position_open`; `position_closed`; `tilt_min`; `tilt_max`. The `entity_id` can be used to reference the entity's attributes with help of the [states](/docs/configuration/templating/#states) template function;"
```

  type: boolean
  default: false
set_position_template:

>   description: "Defines a [template](/topics/templating/) to define the position to be sent to the `set_position_topic` topic. Incoming position value is available for use in the template `{% raw %}{{ position }}{% endraw %}`. Within the template the following variables are available for the calculation of the set-position: __position__, the set-position in percent; __entity_id__, The ID of the entity itself. It can be used to reverence its attribuetes; __position_open__; __position_closed__; __tilt_min__; __tilt_max__. If no template is defined, the position (0-100) will be calculated according to `position_open` and `position_closed` values."

Reviewer: ```suggestion
  description: "Defines a [template](/topics/templating/) to define the position to be sent to the `set_position_topic` topic. Incoming position value is available for use in the template `{% raw %}{{ position }}{% endraw %}`. Within the template the following variables are available: `entity_id`, `position`, the target position in percent; `position_open`; `position_closed`; `tilt_min`; `tilt_max`. The `entity_id` can be used to reference the entity's attributes with help of the [states](/docs/configuration/templating/#states) template function;"
```

  type: integer
  default: 0
tilt_command_template:

>   description: "Defines a [template](/topics/templating/) that can be used to extract the payload for the `tilt_command_topic` topic. Within the template the following variables are available for the calculation of the position: __tilt_position__, the target tilt position in percent; __entity_id__, The ID of the entity itself. It can be used to reverence its attribuetes; __position_open__; __position_closed__; __tilt_min__; __tilt_max__."

Reviewer: ```suggestion
  description: "Defines a [template](/topics/templating/) that can be used to extract the payload for the `tilt_command_topic` topic. Within the template the following variables are available: `entity_id`, `tilt_position`, the target tilt position in percent; `position_open`; `position_closed`; `tilt_min`; `tilt_max`. The `entity_id` can be used to reference the entity's attributes with help of the [states](/docs/configuration/templating/#states) template function;"
```

  type: boolean
  default: "`true` if `tilt_status_topic` is not defined, else `false`"
tilt_status_template:

>   description: "Defines a [template](/topics/templating/) that can be used to extract the payload for the `tilt_status_topic` topic. Within the template the following variables are available for the calculation of the position: __entity_id__, The ID of the entity itself. It can be used to reverence its attribuetes; __position_open__; __position_closed__; __tilt_min__; __tilt_max__."

Reviewer: ```suggestion
  description: "Defines a [template](/topics/templating/) that can be used to extract the payload for the `tilt_status_topic` topic. Within the template the following variables are available: `entity_id`, `position_open`; `position_closed`; `tilt_min`; `tilt_max`. The `entity_id` can be used to reference the entity's attributes with help of the [states](/docs/configuration/templating/#states) template function;"
```

```

{% endraw %}

> ### Full configuration using advanced templating

Reviewer 2: ```suggestion

### Full configuration using advanced templating
```

> The example below shows a full example how to setup a venetian blind, a blind with moveable slats, that tilts with a position change, that does not have a topic for tilt. In the example it takes the blind 6% of the movement for a full rotation of the slats.

Reviewer: ```suggestion
The example below shows a full example of how to setup a venetian blind which has a combined position and tilt topic. The blind in the example has moveable slats which tilt with a position change. In the example it takes the blind 6% of the movement for a full rotation of the slats.
```
Two questions:
- Can the slats really perform a full rotation, or is 6% the movement needed to fully open or close the slats?
- The slats will open or close during the initial 6% of position change, _only after the direction of movement is reversed_, is that right?

> The example below shows a full example of how to setup a venetian blind which has a combined position and tilt topic. The blind in the example has moveable slats which tilt with a position change. In the example it takes the blind 6% of the movement for a full rotation of the slats.

Reviewer 2: ```suggestion
The example below shows a full example of how to set up a venetian blind which has a combined position and tilt topic. The blind in the example has moveable slats which tilt with a position change. In the example, it takes the blind 6% of the movement for a full rotation of the slats.
```

Following variable might be used in `position_template`, `set_position_template`, `tilt_command_template` and `tilt_status_template`, `json_attributes_template` (only `entity_id`).

>  * `entity_id` - The ID of the entity itself. It can be used to reverence its attributes with the help of the [states](/docs/configuration/templating/#states) template function.

Reviewer: ```suggestion
 * `entity_id` - The ID of the entity itself. It can be used to reference its attributes with the help of the [states](/docs/configuration/templating/#states) template function.
```

>  * `entity_id` - The ID of the entity itself. It can be used to reference its attributes with the help of the [states](/docs/configuration/templating/#states) template function.
>  * `position_open`
>  * `position_closed`
>  * `tilt_min`
>  * `tilt_max`

Reviewer 2: ```suggestion
- `entity_id` - The ID of the entity itself. It can be used to reference its attributes with the help of the [states](/docs/configuration/templating/#states) template function.
- `position_open`
- `position_closed`
- `tilt_min`
- `tilt_max`
```

Additionally, it is unclear to me what this list is? It is not described.

{% raw %}

```yaml
# Example configuration.yaml entry
cover:
  - platform: mqtt
    name: "MQTT Cover"
    command_topic: "home-assistant/cover/set"
    state_topic: "home-assistant/cover/state"
    position_topic: "home-assistant/cover/position"
    set_position_topic: "home-assistant/cover/position/set"
    tilt_command_topic: "home-assistant/cover/position/set" # same as `set_position_topic`
    qos: 1
    retain: false
    payload_open:  "open"
    payload_close: "close"
    payload_stop:  "stop"
    state_opening: "open"
    state_closing: "close"
    state_stopped: "stop"
    position_open: 100
    position_closed: 0
    tilt_min: 0
    tilt_max: 6
    tilt_opened_value: 3
    tilt_closed_value: 0
    optimistic: false

>     position_template: |-

Reviewer 2: We should come up with a simpler example, this is really hard to read and understand for the average user.

Reviewer 3: I agree but it is an example which is a real use case, we can add a simpler one so we have two choices:
- replace this one with a simpler example which just shows how to use `entity_id` (and maybe the place for this example is in the forums)
- add a simple example and still leave this one

Anyhow as I see it we should implement `entity_id` and variables for all MQTT components so there should be many examples for this.

Author: I will try to come up wit h a simple example. 
@frenck Please let me know if the complex one shall be kept, or if the usage of variables in templates shall be documented elsewhere? Maybe a description in the templates' documentation would be fine?

      {% if state_attr(entity_id, "current_position") == None %}
        {% set my_json = { "position" : value, "tilt_value" : 0 } %}
      {% else %}
        {% set position = state_attr(entity_id, "current_position") %}
        {% set movement = value | int - position %}
        {% set tilt_percent = (state_attr(entity_id, "current_tilt_position")) %}
        {% set tilt = (tilt_percent/100*(tilt_max-tilt_min)) %}
        {% set tilt_new = tilt + movement %}
        {% if tilt_new > tilt_max %} {% set tilt_new = tilt_max %} {% endif %}
        {% if tilt_new < tilt_min %} {% set tilt_new = tilt_min %} {% endif %}
        {% set my_json = { "position" : value, "tilt_value" : tilt_new, "pos" : position, "tilt" : tilt, "tilt_percent" : tilt_percent, "mov" : movement } %}
      {% endif %}
      {{ my_json }}
    tilt_command_template: |-
      {% set position = state_attr(entity_id, "current_position") %}
      {% set tilt = state_attr(entity_id, "current_tilt_position") %}
      {% set movement = (tilt_position - tilt) / 100 * tilt_max %}
      {{ (position + movement) | int }
```

> {% endraw %}

Reviewer 2: ```suggestion
{% endraw %}

```

{% endraw %}

### Full configuration using `entity_id`- variable in the template

The example below shows an example of how to correct the state of the bline depending if it moved up, or down. 

{% raw %}

```yaml
# Example configuration.yaml entry
cover:
  - platform: mqtt
    name: "MQTT Cover"
    command_topic: "home-assistant/cover/set"
    state_topic: "home-assistant/cover/state"
    position_topic: "home-assistant/cover/position"
    set_position_topic: "home-assistant/cover/position/set"
    payload_open:  "open"
    payload_close: "close"
    payload_stop:  "stop"
    state_opening: "open"
    state_closing: "close"
    state_stopped: "stop"
    optimistic: false
    position_template: |-

>       {% if state_attr(entity_id, "current_position") == None %}
>         {{ value }}
>       {% else %}
>         {% if state_attr(entity_id, "current_position") < (value | int) %}
>           {{ (value | int + 1) }}
>         {% else %}
>           {% if state_attr(entity_id, "current_position") > (value | int) %}
>             {{ (value | int - 1) }}
>           {% else %}
>             {{ value }}
>           {% endif %}
>         {% endif %}
>       {% endif %}

Reviewer 2: ```suggestion
      {% if not state_attr(entity_id, "current_position") %}
        {{ value }}
      {% elif state_attr(entity_id, "current_position") < (value | int) %}
        {{ (value | int + 1) }}
      {% elif state_attr(entity_id, "current_position") > (value | int) %}
        {{ (value | int - 1) }}
      {% else %}
        {{ value }}
      {% endif %}
```

Following variable might be used in `position_template`, `set_position_template`, `tilt_command_template` and `tilt_status_template`, `json_attributes_template` (only `entity_id`).

- `entity_id` - The ID of the entity itself. It can be used to reference its attributes with the help of the [states](/docs/configuration/templating/#states) template function.
- `position_open`
- `position_closed`
- `tilt_min`
- `tilt_max`

{% raw %}

```yaml
# Example configuration.yaml entry
cover:
  - platform: mqtt
    name: "MQTT Cover"
    command_topic: "home-assistant/cover/set"
    state_topic: "home-assistant/cover/state"
    position_topic: "home-assistant/cover/position"
    set_position_topic: "home-assistant/cover/position/set"
    tilt_command_topic: "home-assistant/cover/position/set" # same as `set_position_topic`
    qos: 1
    retain: false
    payload_open:  "open"
    payload_close: "close"
    payload_stop:  "stop"
    state_opening: "open"
    state_closing: "close"
    state_stopped: "stop"
    position_open: 100
    position_closed: 0
    tilt_min: 0
    tilt_max: 6
    tilt_opened_value: 3
    tilt_closed_value: 0
    optimistic: false
    position_template: |-

>       {% if state_attr(entity_id, "current_position") == None %}
>         {% set my_json = { "position" : value, "tilt_value" : 0 } %}
>       {% else %}
>         {% set position = state_attr(entity_id, "current_position") %}
>         {% set movement = value | int - position %}
>         {% set tilt_percent = (state_attr(entity_id, "current_tilt_position")) %}
>         {% set tilt = (tilt_percent/100*(tilt_max-tilt_min)) %}
>         {% set tilt_new = tilt + movement %}
>         {% if tilt_new > tilt_max %} {% set tilt_new = tilt_max %} {% endif %}
>         {% if tilt_new < tilt_min %} {% set tilt_new = tilt_min %} {% endif %}
>         {% set my_json = { "position" : value, "tilt_value" : tilt_new, "pos" : position, "tilt" : tilt, "tilt_percent" : tilt_percent, "mov" : movement } %}
>       {% endif %}
>       {{ my_json }}
>     tilt_command_template: |-
>       {% set position = state_attr(entity_id, "current_position") %}
>       {% set tilt = state_attr(entity_id, "current_tilt_position") %}
>       {% set movement = (tilt_position - tilt) / 100 * tilt_max %}
>       {{ (position + movement) | int }

Reviewer 2: ```suggestion
      {% if not state_attr(entity_id, "current_position") %}
        {
          "position" : value,
          "tilt_value" : 0
        }
      {% else %}
        {% set position = state_attr(entity_id, "current_position") %}
        {% set tilt_percent = (state_attr(entity_id, "current_tilt_position")) %}

        {% set movement = value | int - position %}
        {% set tilt = (tilt_percent / 100 * (tilt_max - tilt_min)) %}
        {% set tilt_value = min(max((tilt + movement), tilt_min), max) %}
 
        {
           "postition": value,
           "pos": position,
           "tilt": tilt,
           "tilt_value": tilt_value,
           "tilt_percent" : tilt_percent,
           "mov" : movement
        }
      {% endif %}
   tilt_command_template: >-
      {% set position = state_attr(entity_id, "current_position") %}
      {% set tilt = state_attr(entity_id, "current_tilt_position") %}
      {% set movement = (tilt_position - tilt) / 100 * tilt_max %}
      {{ position + movement }}
```

I noticed are syntax errors in this template as well. Please make sure they work.

Reviewer 2: @m1ch Friendly ping, could you take a look at these suggestions? Beta cut-off is getting close.
